File: src/main/java/com/hileco/cortex/instructions/bits/BITWISE_NOT.java
package com.hileco.cortex.instructions.bits;

import com.hileco.cortex.context.ProcessContext;
import com.hileco.cortex.context.ProgramContext;
import com.hileco.cortex.context.ProgramZone;
import com.hileco.cortex.instructions.Instruction;
import com.hileco.cortex.instructions.StackParameter;
import lombok.EqualsAndHashCode;

import java.util.Collections;
import java.util.List;

import static com.hileco.cortex.context.ProgramZone.STACK;

@EqualsAndHashCode
public class BITWISE_NOT implements Instruction {

    public static final StackParameter INPUT = new StackParameter("input", 0);

    private static final byte CLEAR = 127;

    @Override
    public void execute(ProcessContext process, ProgramContext program) {
        var stack = program.getStack();
        var pop = stack.pop();
        var result = new byte[pop.length];
        for (var i = 0; i < result.length; i++) {
            result[i] = CLEAR;
            result[i] ^= pop[i];
        }
        stack.push(result);
    }

    @Override
    public List<Integer> getStackAdds() {
        return Collections.singletonList(-1);
    }

    @Override
    public List<ProgramZone> getInstructionModifiers() {
        return Collections.singletonList(STACK);
    }

    @Override
    public List<StackParameter> getStackParameters() {
        return List.of(INPUT);
    }

    @Override
    public String toString() {
        return "BITWISE_NOT";
    }
}

File: src/main/java/com/hileco/cortex/instructions/bits/BITWISE_NOT.kt
package com.hileco.cortex.instructions.bits

import com.hileco.cortex.instructions.Instruction
import com.hileco.cortex.instructions.ProgramException
import com.hileco.cortex.instructions.StackParameter
import com.hileco.cortex.vm.ProgramZone
import com.hileco.cortex.vm.ProgramZone.STACK
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.VirtualMachine
import com.hileco.cortex.vm.symbolic.SymbolicProgramContext
import com.hileco.cortex.vm.symbolic.SymbolicVirtualMachine
import kotlin.experimental.inv

class BITWISE_NOT : Instruction() {
    override val stackAdds: List<Int>
        get() = listOf(-1)

    override val instructionModifiers: List<ProgramZone>
        get() = listOf(STACK)

    override val stackParameters: List<StackParameter>
        get() = listOf(INPUT)

    override fun execute(virtualMachine: VirtualMachine, programContext: ProgramContext) {
        if (programContext.stack.size() < 1) {
            throw ProgramException(ProgramException.Reason.STACK_TOO_FEW_ELEMENTS)
        }
        val element = programContext.stack.pop()
        val result = ByteArray(element.size)
        for (i in result.indices) {
            result[i] = element[i].inv()
        }
        programContext.stack.push(result)
    }

    override fun execute(virtualMachine: SymbolicVirtualMachine, programContext: SymbolicProgramContext) {
        throw UnsupportedOperationException()
    }

    companion object {
        val INPUT = StackParameter("input", 0)
    }
}
